### ISSUE TITLE
Inner enum declarations cause initialization-checker error

### ISSUE THREAD

**Issue by Xavientois** [lampepfl/dotty#14460]
## Compiler version

3.1.2-RC1

## Minimized code

```Scala
class Outer:
    enum MyEnum {
        case Case
    }
```

## Output

```scala
-- Error: tests/init/pos/inner-enum.scala:3:8 --------------------------------------------------------------------------
3 |        case Case
  |        ^
  |        Cannot prove that the value is fully initialized. May only use initialized value as method arguments.
  |
  |        The unsafe promotion may cause the following problem:
  |        Access non-initialized value $values. Calling trace:
  |         -> case Case	[ inner-enum.scala:3 ]
```

## Expectation

When compiling with the `-Ysafe-init` flag this declaration should pass.

**Comment by Xavientois** [comment]
This error appears to originate from the de-sugared code where the `$values` field of the enum's companion object is accessed

**Comment by Xavientois** [comment]
I plan to try to find a fix for this

**Comment by Xavientois** [comment]
This is the de-sugared code generated for the above definition:
```Scala
package dotty.tools.dotc {
     @SourceFile("compiler/src/dotty/tools/dotc/MyEnum.scala") class MyEnum()
        extends
      Object() {
         @Child[(dotty.tools.dotc.MyEnum.MyEnum.Case : MyEnum.this.MyEnum)] sealed
           abstract
          class MyEnum() extends Object(), scala.reflect.Enum {}
         final lazy module val MyEnum: MyEnum.this.MyEnum = new MyEnum.this.MyEnum()
           final module class MyEnum() extends AnyRef(), scala.deriving.Mirror.Sum {
             case val Case: MyEnum.this.MyEnum = this.$new(0, "Case")
               private[this] val $values: Array[MyEnum.this.MyEnum] =
                 Array.apply[MyEnum.this.MyEnum]([this.Case : MyEnum.this.MyEnum])(
                   scala.reflect.ClassTag.apply[MyEnum.this.MyEnum](
                     classOf[MyEnum.this.MyEnum]
                 )
               )
             def values: Array[MyEnum.this.MyEnum] = this.$values.clone()
               def valueOf($name: String): MyEnum.this.MyEnum =
                 $name match
                 {
                     case "Case" => this.Case
                       case _ =>
                         throw
                           new IllegalArgumentException("enum case not found: ".+($name))
                     }
               private[this] def $new(_$ordinal: Int, $name: String): MyEnum.this.MyEnum
                  =
               {
                 final class $anon() extends MyEnum.this.MyEnum(), runtime.EnumValue,
                   scala.deriving.Mirror.Singleton
                    {
                       private def readResolve(): AnyRef =
                           MyEnum.this.MyEnum.this.fromOrdinal(this.ordinal)
                         override def productPrefix: String = $name
                         override def toString(): String = $name
                         override def ordinal: Int = _$ordinal
                       }
                   new
                     MyEnum.this.MyEnum with runtime.EnumValue with
                     scala.deriving.Mirror.Singleton
                      {...}
                     ():MyEnum.this.MyEnum & runtime.EnumValue
                 }
               def fromOrdinal(ordinal: Int): MyEnum.this.MyEnum =
                 try this.$values.apply(ordinal) catch
                   {
                     case _ => throw new NoSuchElementException(ordinal.toString())
                     }
               type MirroredMonoType = MyEnum.this.MyEnum
               def ordinal(x$0: dotty.tools.dotc.MyEnum.MyEnum.MirroredMonoType): Int =
                 x$0.ordinal
             }
       }
     }
```

**Comment by Xavientois** [comment]
Here is the de-sugared code for the above example:
```Scala
package dotty.tools.dotc {
  @SourceFile("compiler/src/dotty/tools/dotc/MyOuter.scala") class MyOuter()
     extends
   Object() {
    @Child[(dotty.tools.dotc.MyOuter.MyEnum.Case : MyOuter.this.MyEnum)] sealed
      abstract
     class MyEnum() extends Object(), scala.reflect.Enum {}
    final lazy module val MyEnum: MyOuter.this.MyEnum =
      new MyOuter.this.MyEnum()
    final module class MyEnum() extends AnyRef(), scala.deriving.Mirror.Sum {
      case val Case: MyOuter.this.MyEnum = this.$new(0, "Case")
      private[this] val $values: Array[MyOuter.this.MyEnum] =
        Array.apply[MyOuter.this.MyEnum]([this.Case : MyOuter.this.MyEnum])(
          scala.reflect.ClassTag.apply[MyOuter.this.MyEnum](
            classOf[MyOuter.this.MyEnum]
          )
        )
      def values: Array[MyOuter.this.MyEnum] = this.$values.clone()
      def valueOf($name: String): MyOuter.this.MyEnum =
        $name match
          {
            case "Case" => this.Case
            case _ =>
              throw
                new IllegalArgumentException("enum case not found: ".+($name))
          }
      private[this] def $new(_$ordinal: Int, $name: String): MyOuter.this.MyEnum
         =
      {
        final class $anon() extends MyOuter.this.MyEnum(), runtime.EnumValue,
          scala.deriving.Mirror.Singleton
         {
          private def readResolve(): AnyRef =
            MyOuter.this.MyEnum.this.fromOrdinal(this.ordinal)
          override def productPrefix: String = $name
          override def toString(): String = $name
          override def ordinal: Int = _$ordinal
        }
        new
          MyOuter.this.MyEnum with runtime.EnumValue with
            scala.deriving.Mirror.Singleton
           {...}
        ():MyOuter.this.MyEnum & runtime.EnumValue
      }
      def fromOrdinal(ordinal: Int): MyOuter.this.MyEnum =
        try this.$values.apply(ordinal) catch
          {
            case _ => throw new NoSuchElementException(ordinal.toString())
          }
      type MirroredMonoType = MyOuter.this.MyEnum
      def ordinal(x$0: dotty.tools.dotc.MyOuter.MyEnum.MirroredMonoType): Int =
        x$0.ordinal
    }
  }
}
```

Further investigation indicates that this error comes from the definition of `readResolve` in the class `$anon` defined within `$new`.  This method is not explicitly called anywhere in the generated code, but `readResolve` is used for serialization.

**Comment by liufengyun** [comment]
This is not yet fixed, will be in #14751.
